File: src/main/java/kiinse/plugins/api/darkwaterapi/commands/manager/annotation/Command.java
package kiinse.plugins.api.darkwaterapi.commands.manager.annotation;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface Command {

    /**
     * The command that the player must execute. For example /darkwater reload. Should start with a slash
     * @return Command
     */
    String command();

    /**
     * Number of received parameters. Default - 0
     * @return Number of parameters
     */
    int parameters() default 0;

    /**
     * Allows you to disable restrictions on the number of parameters in a command. Disabled by default
     * @return Enabled or Disabled
     */
    boolean overrideParameterLimit() default false;

    /**
     * Allows you to disable the use of the command by non-players (for example, via the console). Disabled by default
     * @return Enabled or Disabled
     */
    boolean disallowNonPlayer() default false;

    /**
     * Required rights to use the command. Default is empty
     * @return Permissions to use the command
     */
    String permission() default "";

}

File: api/src/main/kotlin/kiinse/me/plugins/darkwaterapi/api/commands/Command.kt
// MIT License
//
// Copyright (c) 2022 kiinse
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
package kiinse.me.plugins.darkwaterapi.api.commands

@Suppress("UNUSED")
@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.CLASS, AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
annotation class Command(
        /**
         * The command that the player must execute. For example /darkwater reload. Should start with a slash
         *
         * @return Command
         */
        val command: String,
        /**
         * Allows you to disable the use of the command by non-players (for example, via the console). Disabled by default
         *
         * @return Enabled or Disabled
         */
        val disallowNonPlayer: Boolean = false,
        /**
         * Required rights to use the command. Default is empty
         *
         * @return Permissions to use the command
         */
        val permission: String = "")